app/data_model/database.py

```diff
@@ -73,68 +66,65 @@ def __repr__(self):
-def test_connection(connection, branch):
-    if branch:
-        # "branch" refers to a sub-connection of a connection,
-        # we don't want to bother testing on these.
-        return
-
-    save_should_close_with_result = connection.should_close_with_result
-    connection.should_close_with_result = False
-    try:
-        # Run a SELECT 1 to test the database connection
-        connection.scalar(select([1]))
-    except exc.DBAPIError:
-        try:
-            # try it again, this will recreate a stale or broke connection
-            connection.scalar(select([1]))
-        except exc.DBAPIError:
-            # However if we get a database err again, forcibly close the connection
-            connection.close()
-
-    finally:
-        # restore "close with result"
-        connection.should_close_with_result = save_should_close_with_result
-
-
-@contextmanager
-def commit_or_rollback(database_session):
-    try:
-        yield database_session
-        database_session.commit()
-    except SQLAlchemyError:
-        database_session.rollback()
-        raise
+class Database:
+
+    def __init__(self, database_url, database_setup_retry_count, database_setup_retry_delay):
+        self.db_session, self.engine = self._create_session_and_engine_with_retry(
+            database_url,
+            database_setup_retry_count,
+            database_setup_retry_delay)
+
+    @staticmethod
+    def _create_session_and_engine(database_url):
+        logger.info("setting up database...")
+        logger.debug("creating database engine")
+        eng = create_engine(database_url, convert_unicode=True)
+        logger.debug("creating database session")
+        session = scoped_session(sessionmaker(autocommit=False, autoflush=False, bind=eng))
+        base.query = session.query_property()
+        logger.debug("creating database and tables")
+        base.metadata.create_all(eng)
+        logger.info('database setup complete.')
+        return session, eng
+
+    def _create_session_and_engine_with_retry(self, database_url, database_setup_retry_count, database_setup_retry_delay):
+        last_exception = None
+
+        for i in range(database_setup_retry_count):
+            try:
+                return self._create_session_and_engine(database_url)
+            except Exception as e:  # pylint: disable=broad-except
+                # Accept catching of Exception here as we re-throw it below if the retry fails.
+                last_exception = e
+                logger.error('error setting up database', exc_info=e, attempt=i)
+                time.sleep(database_setup_retry_delay)
+        else:  # pylint: disable=useless-else-on-loop
+            logger.error('failed to setup database, aborting', exc_info=last_exception)
+            raise TimeoutError('failed to setup database') from last_exception
+
+    def remove(self):
+        self.db_session.remove()
+
+    # @event.listens_for(engine, "engine_connect")
```

Comment:
This was un-intentionally left commented. However I'm pretty sure re-connection doesn't work anyway.